modules/auxiliary/admin/ldap/change_password.rb

```diff
@@ -0,0 +1,159 @@
+          ['URL', 'https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-adts/6e803168-f140-4d23-b2d3-c3a8ab5917d2'],
+        ],
+        'License' => MSF_LICENSE,
+        'Actions' => [
+          ['RESET', { 'Description' => "Reset a target user's password, having permissions over their account" }],
+          ['CHANGE', { 'Description' => "Change the user's password, knowing the existing password" }]
+        ],
+        'DefaultAction' => 'RESET',
+        'Notes' => {
+          'Stability' => [],
+          'SideEffects' => [ IOC_IN_LOGS ],
+          'Reliability' => []
+        }
+      )
+    )
+
+    register_options([
+      OptString.new('TARGET_USER', [false, 'The user to reset the password of.'], conditions: ['ACTION', 'in', %w[RESET]]),
+      OptString.new('NEW_PASSWORD', [ true, 'The new password to set for the user' ])
+    ])
+  end
+
+  def fail_with_ldap_error(message)
+    ldap_result = @ldap.get_operation_result.table
+    return if ldap_result[:code] == 0
+
+    print_error(message)
+    if ldap_result[:code] == 19
+      extra_error = ''
+      if action.name == 'CHANGE' && !datastore['SESSION'].blank?
+        # If you're already in a session, you could provide the wrong password, and you get this error
+        extra_error = ' or incorrect current password'
+      end
+
+      error = "The password changed failed, likely due to a password policy violation (e.g. not sufficiently complex, matching previous password, or changing the password too often)#{extra_error}"
+      fail_with(Failure::NotFound, error)
+    else
+      validate_query_result!(ldap_result)
+    end
+  end
+
+  def ldap_get(filter, attributes: [])
+    raw_obj = @ldap.search(base: @base_dn, filter: filter, attributes: attributes)&.first
+    return nil unless raw_obj
+
+    obj = {}
+
+    obj['dn'] = raw_obj['dn'].first.to_s
+    unless raw_obj['sAMAccountName'].empty?
+      obj['sAMAccountName'] = raw_obj['sAMAccountName'].first.to_s
+    end
+
+    unless raw_obj['ObjectSid'].empty?
+      obj['ObjectSid'] = Rex::Proto::MsDtyp::MsDtypSid.read(raw_obj['ObjectSid'].first)
+    end
+
+    obj
+  end
+
+  def run
+    if action.name == 'CHANGE'
+      fail_with(Failure::BadConfig, 'Must set USERNAME when changing password') if datastore['USERNAME'].blank?
+      fail_with(Failure::BadConfig, 'Must set PASSWORD when changing password') if datastore['PASSWORD'].blank?
+    end
+    if session.blank? && datastore['USERNAME'].blank?
+      print_warning('Connecting with an anonymous bind')
+    end
+    ldap_connect do |ldap|
+      validate_bind_success!(ldap)
+
+      if (@base_dn = datastore['BASE_DN'])
+        print_status("User-specified base DN: #{@base_dn}")
+      else
+        print_status('Discovering base DN automatically')
+
+        if (@base_dn = ldap.base_dn)
+          print_status("#{ldap.peerinfo} Discovered base DN: #{@base_dn}")
+        else
+          print_warning("Couldn't discover base DN!")
```

Comment:
In my understanding, `@base_dn` will be `nil` if it is not provided in `datastore['BASE_DN']` or if it has not been discovered automatically. This instance variable is used later in `#ldap_get` in the LDAP search call. Is it an issue (I've not tested)?